### ISSUE TITLE
No Join Type in MergeJoin?

### ISSUE THREAD

**Issue by gjvanvuuren** [etlbox/etlbox#73]
Hi,

I'm using using MergeJoin with the intent of **Left** joining two data sets. But I noticed that MergeJoin by default performs an **Inner** join. In SSIS at least, the MergeJoin transformation allows the user to choose between different Join types.

How would I go about to do a Left join with EtlBox?

Regards,
Gideon

**Comment by roadrunnerlenny** [comment]
Hi,

Thanks for your interest in ETLBox. 

Just to make sure I understand your question correctly: the MergeJoin combines data from two inputs into one output. So basically you are joining two sets of incoming data which are uneven? e.g. one input has three rows and one has two rows, so you want to have the last rows of input one processed by the MergeJoin?

**Comment by gjvanvuuren** [comment]
Hi,

 

Yes, I have two uneven inputs, or sets, so using your example and the left join analogy, all of input one will be passed through, along with only the matching of input two. The MergeJoin constructor parameter mergeJoinFunc then determines which of the fields of input one and the (matched) input two are passed through.

 

In fact, I assumed that’s what the mergeJoinFunc function is for – to implement left join, right join, or outer join logic. But it turns out only the inner joined set of values are passed to the mergeJoinFunc delegate.

 

By the way, this is a great project – a.f.a.i.k. the first viable .NET library for code driven ETL / ELT. I’m seriously considering gradually phasing out SSIS.

 

Thanks!

Gideon

**Comment by roadrunnerlenny** [comment]
Hi @gjvanvuuren ,

that's actually a good point you made. The question is how the MergeJoin can be enhanced to support other join types regarding the programming interface.

I just compared the implementation in SSIS - the MergeJoin is a little bit differently there regarding the configuration. In ETLBox, you basically can define any function that decides if and how a data row would be passed (e.g. you could return null here, or you could do any assignments between the properties you want). On the other hand, in SSIS you basically define which properties are used for joining and which properties are used to "enrich the output".

So the question is: Could you me an example code how you would expect the MergeJoin to work when you want to use it for a left/right or full join? So that I get the basic idea how you would like to call the MergeJoin? That would help me to get the implementation started, or at least define the tests with the expected result.

**Comment by gjvanvuuren** [comment]
Hi Andreas,

 

I’ve written a simple SQL statement, and the equivalent EtlBox flow to illustrate the problem.

 

SQL:

 

WITH 

FirstName AS 

(

    SELECT 1 AS ID, 'Marilyn' AS FirstName UNION

    SELECT 2 AS ID, 'James' AS FirstName UNION

    SELECT 3 AS ID, 'Elvis' AS FirstName

),

 

LastName AS

(

    SELECT 1 AS ID, 'Monroe' AS LastName UNION

    SELECT 3 AS ID, 'Presley' AS LastName

)

 

SELECT 

    fn.ID,

    fn.FirstName,

    ln.LastName,

    fn.FirstName + ' ' + ln.LastName AS FullName

FROM

    FirstName AS fn

LEFT JOIN

    LastName AS ln

ON fn.ID = ln.ID

 

 

ETLBox:

 

class Program

{

    static void Main(string[] args)

    {

        IList<dynamic> FirstName = new[]

        {

            new { ID = 1, FirstName = "Marilyn" },

            new { ID = 2, FirstName = "James" },

            new { ID = 3, FirstName = "Elvis" }

        }.Select

        (

            x =>

            {

                dynamic fn = new ExpandoObject();

                fn.ID = x.ID;

                fn.FirstName = x.FirstName;

                return fn;

            }

        ).ToList();

 

        IList<dynamic> LastName = new[]

        {

            new { ID = 1, LastName = "Monroe" },

            new { ID = 3, LastName = "Presley" }

        }.Select

        (

            x =>

            {

                dynamic ln = new ExpandoObject();

                ln.ID = x.ID;

                ln.LastName = x.LastName;

                return ln;

            }

        ).ToList();

 

        var firstNameSource = new MemorySource<dynamic>(FirstName);

        var lastNameSource = new MemorySource<dynamic>(LastName);

 

        var join = new MergeJoin<dynamic, dynamic, dynamic>

        (

            // SQL-equivalent Left Join

            (left, right) =>

            {

                dynamic dleft = left as ExpandoObject;

                dynamic dright = right as ExpandoObject;

 

                dynamic dout = new ExpandoObject();

 

                // return everything from the left...

                dout.ID = dleft.ID;

                dout.FirstName = dleft.FirstName;

 

                // ... but only matching from the right ...

                if (dleft.ID == dright.ID)

                {

                    dout.LastName = dright.LastName;

                    dout.FullName = dleft.FirstName + " " + dright.LastName;

                }

                // ... right defaults to null

                else

                {

                    dout.LastName = null;

                    dout.FullName = null;

                }

 

                return dout;

            }

        );

 

        var fullNameDest = new MemoryDestination<dynamic>();

 

        firstNameSource.LinkTo(join.Target1);

        lastNameSource.LinkTo(join.Target2);

 

        join.LinkTo<dynamic>(fullNameDest);

 

        firstNameSource.Execute();

        lastNameSource.Execute();

 

        fullNameDest.Wait();

    }

}

 

Now it’s easy to debug and see that ETLBox only passes in 2 records to the Merge function. From the FirstName - ID’s 1, 2; from the LastName, ID’s 1, 3. ID 3 from the FirstName table is ignored.

 

Regards,

Gideon

**Comment by gjvanvuuren** [comment]
Hi Andreas,

I’ve written a simple SQL statement, and the equivalent EtlBox flow to illustrate the problem.

**SQL:**

```sql
WITH 
FirstName AS 
(
    SELECT 1 AS ID, 'Marilyn' AS FirstName UNION
    SELECT 2 AS ID, 'James' AS FirstName UNION
    SELECT 3 AS ID, 'Elvis' AS FirstName
),

LastName AS
(
    SELECT 1 AS ID, 'Monroe' AS LastName UNION
    SELECT 3 AS ID, 'Presley' AS LastName
)

SELECT 
    fn.ID,
    fn.FirstName,
    ln.LastName,
    fn.FirstName + ' ' + ln.LastName AS FullName
FROM
    FirstName AS fn
LEFT JOIN
    LastName AS ln
ON fn.ID = ln.ID
```

**ETLBox:**

```csharp
class Program
{
    static void Main(string[] args)
    {
        IList<dynamic> FirstName = new[]
        {
            new { ID = 1, FirstName = "Marilyn" },
            new { ID = 2, FirstName = "James" },
            new { ID = 3, FirstName = "Elvis" }
        }.Select
        (
            x =>
            {
                dynamic fn = new ExpandoObject();
                fn.ID = x.ID;
                fn.FirstName = x.FirstName;
                return fn;
            }
        ).ToList();

        IList<dynamic> LastName = new[]
        {
            new { ID = 1, LastName = "Monroe" },
            new { ID = 3, LastName = "Presley" }
        }.Select
        (
            x =>
            {
                dynamic ln = new ExpandoObject();
                ln.ID = x.ID;
                ln.LastName = x.LastName;
                return ln;
            }
        ).ToList();

        var firstNameSource = new MemorySource<dynamic>(FirstName);
        var lastNameSource = new MemorySource<dynamic>(LastName);

        var join = new MergeJoin<dynamic, dynamic, dynamic>
        (
            // SQL-equivalent Left Join
            (left, right) =>
            {
                dynamic dleft = left as ExpandoObject;
                dynamic dright = right as ExpandoObject;

                dynamic dout = new ExpandoObject();

                // return everything from the left...
                dout.ID = dleft.ID;
                dout.FirstName = dleft.FirstName;

                // ... but only matching from the right ...
                if (dleft.ID == dright.ID)
                {
                    dout.LastName = dright.LastName;
                    dout.FullName = dleft.FirstName + " " + dright.LastName;
                }
                // ... right defaults to null
                else
                {
                    dout.LastName = null;
                    dout.FullName = null;
                }

                return dout;
            }
        );

        var fullNameDest = new MemoryDestination<dynamic>();

        firstNameSource.LinkTo(join.Target1);
        lastNameSource.LinkTo(join.Target2);

        join.LinkTo<dynamic>(fullNameDest);

        firstNameSource.Execute();
        lastNameSource.Execute();

        fullNameDest.Wait();
    }
}
```

Now it’s easy to debug and see that ETLBox only passes in 2 records to the Merge function. From the FirstName - ID’s 1, 2; from the LastName, ID’s 1, 3. ID 3 from the FirstName table is ignored.

Regards,
Gideon

**Comment by gjvanvuuren** [comment]
I unintentionally closed this issue - reopening.

**Comment by roadrunnerlenny** [comment]
Hi @gjvanvuuren ,
great, that's actually quite a good example. Thank you for taking the time to prepare this.

Now I got a much better understanding of the problem. 
So the current merge join is basically taking data from both inputs, and the MergeFunc just describes how to join them. This can work well if data from both inputs have the same size, and are already sorted. 

This doesn't work if the data is "unevenly" spread. So in your left join scenario, you would want to join the data on the id,  but you actually receive ID 2 and 3 in the same moment. 

So you need a MergeJoin where you define the following:
- Some kind of join predicate which describes when two objects match (in the example that would be if the Ids are equally)
- Optionally: An [MergeJoinKey] attribute to easily mark the properties that should be used for the join instead providing the match condition
- A function that is executed if the ids match
- A function that is executed if the ids don't match 

For performance reason, the MergeJoin still would need sorted input on both sides. It then basically can receive the data and always decide which record  "to keep" until it receives the matching record. 

Is this what you had in mind?
It does sound possible to implement this, but let me first look into this. The MergeJoin right now is a non blocking transformation, which means that it actually never keeps more data in memory than the current record that is processed (plus some in the buffer queue to be faster). The changes above shouldn't change this, but I need to think this through first.

Did you have a look at the LookupTransformation (https://etlbox.net/articles/dataflow/dataflow_transformations.html#lookuptransformation)? It would actually solve your example code here, as it generally works like a left join. It could be compared with the LookupTransformation in SSIS. The only downside here is that one part of the data needs to be loaded into the memory in full (there is currently no "no cache" or "partial cache" option like in SSIS)

**Comment by gjvanvuuren** [comment]
Hi Andreas,

thanks for the response and your suggestions. 

I don't think MergJoin has to be changed much - the requirement to have sorted input is standard practice. The SQL optimizer also sorts the tables based on the join criteria. SSIS also requires sorted input.

The Merge Function can be used as is, the user can define their own join logic using the existing design. The problem as I see it, lies with the use of the Tuple in the definition of the Transformation function in the MergeJoinFunc setter:

```csharp
Transformation.TransformationFunc = new Func<Tuple<TInput1, TInput2>, TOutput>(tuple => _mergeJoinFunc.Invoke(tuple.Item1, tuple.Item2));
```

You define a non-nullable Tuple as input. I think the Tuple parameters should be made nullable, something like:

```csharp
Transformation.TransformationFunc = new Func<Tuple<TInput1?, TInput2?>, TOutput>(tuple => _mergeJoinFunc.Invoke(tuple.Item1, tuple.Item2));
```

Null values are inevitable in joining tables.

The above statement doesn't compile, as the compiler cannot cater for reference types and value types. So I'm not sure yet how to do this in practice, but maybe this is something that you can look at?

Regards,
Gideon

**Comment by roadrunnerlenny** [comment]
Hi @gjvanvuuren ,

that is a good starting point and good thoughts.

I am still struggling with the uneven inputs - one side of the join can have 3 rows, the other one 2 rows. Still you would expect 3 rows in the output. 
The MergeBlock from the TPL Dataflow library which I use under the hood would just simple merge rows from both sides - if one side is delivering a row and the other not it would not process this (last) record.

Anyways, currently I am refactoring the whole linking and object initializing, in order to add new features. I already refactored most of the transformations, the MergeJoin is still outstanding - I will have a deeper look into this the next couple of days. I am confident that I improve the functionality here. 

Regarding the transformation `Func`: Even though the function could be something like `Func<TInput1, TInput2, TOutput>` (basically removing the tuple, which will allow nulls), there still must be some kind of join condition, otherwise there is no chance to know when to continue with the next record or when to pass a null into the function. But you are right, the transformation function can still be used, and along with another function/predicate or some attributes for the join condition this could work. 

I'll keep you updated when I made progress here.

**Comment by gjvanvuuren** [comment]
Hi Andreas,

thanks again for the feedback. I'm looking forward to all the upcoming changes!

Best Regards,
Gideon

**Comment by andreaslennartz** [comment]
Hi @gjvanvuuren ,

a quick update on this: It took me longer than expected, but now I came up with a first working solution.
Initially, I wanted to add a `BothMatchFunc<TInput1,TInput2,bool>` as additionally input parameter that basically described when two records are equal and should be joined or not. 
But it turned out that this approach woudn't work without keeping all data in memory. Otherwise, the MergeJoin wouldn't able to decide which record to keep and which one to join with a "null" without knowing the records that will come. 

Turns out that the missing information was the sort order of the input, respective the order of a record compared to another. 
So I added a `ComparisonFunc<TInput1, TInput2, int>` which returns an int value: 
https://github.com/etlbox/etlbox/blob/8b709bcc0bad5236b3fe73b511c328d67a17434b/ETLBox/src/Toolbox/DataFlow/MergeJoin.cs#L57-L64

If the ComparisonFunc is defined, records are compared regarding their sort order and only joined if they match. 
It returns 0 if both records match and should be joined. It returns a value little than 0 if the record of the left input is in the sort order before the record of the right input. It returns a value greater than 0 if the record for the right input is in the order before the record from the left input.

Here an example how this would look like 

```
            //Arrange
            MemorySource<MySimpleRow> source1 = new MemorySource<MySimpleRow>();
            source1.DataAsList.Add(new MySimpleRow() { Col1 = 1, Col2 = "1" });            
            source1.DataAsList.Add(new MySimpleRow() { Col1 = 3, Col2 = "3" });
            source1.DataAsList.Add(new MySimpleRow() { Col1 = 4, Col2 = "4" });            
            MemorySource<MySimpleRow> source2 = new MemorySource<MySimpleRow>();
            source2.DataAsList.Add(new MySimpleRow() { Col1 = 2, Col2 = "2" });
            source2.DataAsList.Add(new MySimpleRow() { Col1 = 3, Col2 = "3" });

            MemoryDestination<MySimpleRow> dest = new MemoryDestination<MySimpleRow>();

            //Act
            MergeJoin<MySimpleRow, MySimpleRow, MySimpleRow> join = new MergeJoin<MySimpleRow, MySimpleRow, MySimpleRow>();
            join.ComparisonFunc = (inputRow1, inputRow2) =>
            {
                if (inputRow1.Col1 == inputRow2.Col1)
                    return 0;
                else if (inputRow1.Col1 < inputRow2.Col1)
                    return -1;
                else if (inputRow1.Col1 > inputRow2.Col1)
                    return 1;
                else
                    throw new System.Exception("No valid result");
            };
            join.MergeJoinFunc =
                (inputRow1, inputRow2) =>
                {
                    return new MySimpleRow()
                    {
                        Col1 = inputRow1?.Col1 ?? inputRow2.Col1,
                        Col2 = (inputRow1?.Col2 ?? "X") + " " + (inputRow2?.Col2 ?? "X")
                    };
                };
            source1.LinkTo(join.LeftInput);
            source2.LinkTo(join.RightInput);
            join.LinkTo(dest);
            source1.Execute();
            source2.Execute();
            dest.Wait();

            //Assert
            Assert.Collection(dest.Data,
                r => Assert.True(r.Col1 == 1 && r.Col2 == "1 X"),
                r => Assert.True(r.Col1 == 2 && r.Col2 == "X 2"),
                r => Assert.True(r.Col1 == 3 && r.Col2 == "3 3"),
                r => Assert.True(r.Col1 == 4 && r.Col2 == "4 X"),  
                );
```

So this is the solution I am thinking about to put into the next release. It would be great to hear your thoughts, or if you think I did miss something here.

**Comment by andreaslennartz** [comment]
Hi @gjvanvuuren ,
just released version 2.1.2 which contains the improved MergeJoin:  https://www.nuget.org/packages/ETLBox/
It does now accept a ComparisonFunc as decribed before.
 I updated the docu as well: https://etlbox.net/articles/transformations/broadcast_merge_join.html
